TidepoolServiceKit/TidepoolService.swift

```diff
@@ -427,63 +403,39 @@ extension TidepoolService: RemoteDataService {
 
-        Task {
-            do {
-                let result = try await createData(stored.compactMap { $0.datum(for: userId, hostIdentifier: hostIdentifier, hostVersion: hostVersion) })
-                completion(.success(result))
-            } catch {
-                completion(.failure(error))
-            }
-        }
+        let _ = try await createData(stored.compactMap { $0.datum(for: userId, hostIdentifier: hostIdentifier, hostVersion: hostVersion) })
     }
 
     public var pumpDataEventLimit: Int? { return 1000 }
 
-    public func uploadPumpEventData(_ stored: [PersistedPumpEvent], completion: @escaping (_ result: Result<Bool, Error>) -> Void) {
+    public func uploadPumpEventData(_ stored: [PersistedPumpEvent]) async throws {
         guard let userId = userId, let hostIdentifier = hostIdentifier, let hostVersion = hostVersion else {
-            completion(.failure(TidepoolServiceError.configuration))
-            return
+            throw TidepoolServiceError.configuration
         }
 
-        Task {
-            do {
-                let result = try await createData(stored.flatMap { $0.data(for: userId, hostIdentifier: hostIdentifier, hostVersion: hostVersion) })
-                completion(.success(result))
-            } catch {
-                completion(.failure(error))
-            }
-        }
+        let _ = try await createData(stored.flatMap { $0.data(for: userId, hostIdentifier: hostIdentifier, hostVersion: hostVersion) })
     }
 
     public var settingsDataLimit: Int? { return 400 }  // Each can be up to 2.5K bytes of serialized JSON, target ~1M or less
 
-    public func uploadSettingsData(_ stored: [StoredSettings], completion: @escaping (_ result: Result<Bool, Error>) -> Void) {
+    public func uploadSettingsData(_ stored: [StoredSettings]) async throws {
         guard let userId = userId, let hostIdentifier = hostIdentifier, let hostVersion = hostVersion else {
-            completion(.failure(TidepoolServiceError.configuration))
-            return
+            throw TidepoolServiceError.configuration
         }
 
         let (created, updated, lastControllerSettingsDatum, lastCGMSettingsDatum, lastPumpSettingsDatum) = calculateSettingsData(stored, for: userId, hostIdentifier: hostIdentifier, hostVersion: hostVersion)
 
-        Task {
-            do {
-                let createdUploaded = try await createData(created)
-                let updatedUploaded = try await updateData(updated)
-                self.lastControllerSettingsDatum = lastControllerSettingsDatum
-                self.lastCGMSettingsDatum = lastCGMSettingsDatum
-                self.lastPumpSettingsDatum = lastPumpSettingsDatum
-                self.completeUpdate()
-                completion(.success(createdUploaded || updatedUploaded))
-            } catch {
-                completion(.failure(error))
-            }
-        }
+        let _ = try await createData(created)
+        let _ = try await updateData(updated)
```

Comment:
Similar to above. Another possible use of `TaskGroup`?

```suggestion
        await withThrowingTaskGroup(of: Void.self) { group in
            group.addTask { try await createData(created) }
            group.addTask { try await updateData(updated) }
            await group.waitForAll()
        }
```

Reply:
Serialization is intended here.